### ISSUE TITLE
[HOLD for payment 2024-10-14] API returns 407 Session expired but the app does not send you to the login screen.

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#48750]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 
**Reproducible in staging?:** Needs reproduction
**Reproducible in production?:** Needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @coleaeason 
**Slack conversation:** https://expensify.slack.com/archives/C05LX9D6E07/p1725564118248469

## Action Performed:
1. Go to staging.new.expensify.com

## Expected Result:
No loading skeletons or no other issues
## Actual Result:
API returns 407 Session expired but the app does not send you to the login screen.

## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

![Screenshot 2024-09-05 at 3 20 28 PM](https://github.com/user-attachments/assets/ba24e089-fa09-46e3-a2dc-78eebb183d6d)

![Screenshot 2024-09-05 at 3 21 01 PM](https://github.com/user-attachments/assets/787b19c8-2e2b-4919-8823-8dceeb8512a8)

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Issue Owner</summary>Current Issue Owner: @eVoloshchak</details>

**Comment by MelvinBot** [comment]
This has been labelled "Needs Reproduction". Follow the steps here: https://stackoverflowteams.com/c/expensify/questions/16989

**Comment by dangrous** [comment]
Just calling out that the Slack thread(s) have some initial investigation into this one - https://expensify.slack.com/archives/C05LX9D6E07/p1725564162189379 and https://expensify.slack.com/archives/C05LX9D6E07/p1725564118248469 . The only thing I can think of is that the token returned from RestoreAuthToken is not a "normal" token but it certainly appears to be.

I believe there IS an issue with not returning to the default account if a delegated access token expires (which would be nice to fix) but this seems separate, it happened AFTER @coleaeason returned to his actual account.

**Comment by jasperhuangg** [comment]
Based on the linked threads it seems like:
- Something is wrong with the front-end that is preventing us from calling Authenticate correctly when we get the 407 response.
- This isn't entirely related to the issue, but the back-end is returning the expected authToken to the front-end when we switch from co-pilot to normal account.

@puneetlath @dangrous Is there a way that we can reliably reproduce this issue? I'm having trouble doing so by hardcoding our API to return the 407 response.

**Comment by dangrous** [comment]
hm, I'm not sure - you could maybe:
* Log into an account as a copilot
* Adjust https://github.com/Expensify/Auth/blob/5b1b0990bc92b6a345e28df466d8548dc672fdc5/auth/lib/AuthToken.cpp#L1021 to have a very short expiration time
* Return to the original account

And then see what happens?

Otherwise, in terms of actual use I think it's just log into copilot and back to original and just wait, which I realize is not helpful.

**Comment by jasperhuangg** [comment]
@dangrous Thanks for those reproduction steps, I'll try them out and see if that yields any different results.

**Comment by jasperhuangg** [comment]
Hmm I'm unable to reproduce the issue following those reproduction steps, here's what I did:
1. Create two accounts, one main account and one copilot account.
2. Using the main account, set the copilot as the delegate via OldDot.
3. Sign into the copilot account and copilot into the main account.
4. Update the authToken created [here](https://github.com/Expensify/Auth/blob/5b1b0990bc92b6a345e28df466d8548dc672fdc5/auth/lib/AuthToken.cpp#L1021) to expire in a very short amount of time.
5. Exit copilot access.
6. Switch to NewDot.
7. Nothing happens?

Gonna try to look at logs now since I'm unable to reproduce.

**Comment by jasperhuangg** [comment]
[These `ReconnectApp` requests](https://staging.expensify.com/_devportal/tools/logSearch/#sort=asc&size=10000&query=blob%3A%22407%22%20AND%20blob%3A%22ReconnectApp%22%20AND%20email%3A%22cole%40expensify.com%22%20AND%20timestamp%3A%5B2024-09-05T19%3A00%20TO%202024-09-05T20%3A00%5D) are from around the same time that this issue was reported.

They all seem to be throwing [here](https://github.com/Expensify/Auth/blob/89e548c759ecd0c6739efb03aa0adb36f9c412cc/auth/lib/AuthToken.cpp#L1209-L1212), which indicates that the token we were sending with `ReconnectApp` was either expired or not associated with an account.

I've omitted screenshots since this is a public repo.

**Comment by jasperhuangg** [comment]
@dangrous Added some logs and it seems we never get to pushing `Authenticate` onto the queue after getting a 407 response (this is from me hard-coding the user's authToken to expire in 2 seconds).

It seems the reason we don't reauthenticate is because we're failing the requireParameters check [here](https://github.com/Expensify/App/blob/51a64cfe90a9314b51a5f3de00c945aeeea4ee1b/src/libs/requireParameters.ts#L8-L31). Namely it appears from my logs that we're missing the partnerUserID and partnerUserSecret parameters.

<img width="775" alt="Screenshot 2024-09-14 at 11 28 46 AM" src="https://github.com/user-attachments/assets/508365c0-8501-4628-9b16-feff25a36f79">

Maybe we're failing to set these parameters again when the user restores their session when coming back from copilot access? Is there an easy way to check that?

**Comment by jasperhuangg** [comment]
Not overdue just waiting on a response above cc @dangrous

**Comment by jasperhuangg** [comment]
bump @dangrous on above 🙇

**Comment by dangrous** [comment]
Oh hm interesting - maybe we need to add a call to [`createTemporaryLogin`](https://github.com/Expensify/Web-Expensify/blob/84d5a62795dfe50c4d953d6d86c5113a609399e1/lib/UserAPI.php#L3909) inside of [`UserAPI::delegateDisconnect`](https://github.com/Expensify/Web-Expensify/blob/84d5a62795dfe50c4d953d6d86c5113a609399e1/lib/UserAPI.php#L749) if the request is coming from newDot - so that we appropriately send the `partnerUserID` and `partnerUserSecret`? It looks like you're right and those should be being sent but are not.

**Comment by jasperhuangg** [comment]
@dangrous Thanks for the context, I'll look into returning those with the response.

**Comment by jasperhuangg** [comment]
After some more investigation it seems that the real issue is that we're setting the locally stored autoGeneratedLogin/Password to `null` at some point. I still think we should send back a new temporary login, but we also need to figure out where it's being cleared out.

**Comment by jasperhuangg** [comment]
Noticed some other issues during the delegate access flow that were related and needed to be addressed, so I pushed some more changes and a new PR.

**Comment by jasperhuangg** [comment]
@eVoloshchak was pullerbear'd on the linked PR

**Comment by jasperhuangg** [comment]
PRs still under review, @eVoloshchak please prioritize this review!

**Comment by jasperhuangg** [comment]
PR was merged

**Comment by puneetlath** [comment]
@eVoloshchak can you complete the checklist please?

**Comment by puneetlath** [comment]
@eVoloshchak bump on the checklist so that we can pay you out and close this issue. Thanks!

**Comment by eVoloshchak** [comment]
@puneetlath, I'm not owed any payment here, I didn't review https://github.com/Expensify/App/pull/49375 since it wasn't possible for C+ (need someone in ring1, more details [here](https://github.com/Expensify/App/pull/49375#issuecomment-2386453050))

**Comment by puneetlath** [comment]
Ah ok, I see. Thanks @eVoloshchak. @jasperhuangg do you think we need a regression test here. Is there a way Applause could manually test this?

**Comment by jasperhuangg** [comment]
@puneetlath Absolutely, I can create an issue to get a regression test added to the testing suite.

**Comment by puneetlath** [comment]
Ok that'd be great thanks! Feel free to close this issue once you have.
